Unprivileged handles

wferi opened the pull request:
So for the controversial part: here is my followup to #97. It isn't exactly what was suggested there; I was looking for a sound but non intrusive solution, preferably without global state. This is what I was left with, how do you like it?

The doxyxml change was needed because I didn't overdo the documentation part. Consider that more as an independent question.

wferi commented:
Maybe these are more meaningful names...

chrissie-c commented:
Maybe make the new parameter into a 'flags' bitmap and one of the flags is KNET_HANDLE_FLAG_TEST or KNET_HANDLE_FLAG_UNPRIVILEGED ?

wferi commented:
I like the idea! We surely won't have that many privilege levels, but new flags may come in handy.
So `knet_handle_new_ex(..., unsigned int flags)`?

fabbione commented:
Hey guys, I am not particularly excited about adding a whole new api call just for testing.

If we are a looking at a short term solution to allow test suite to run unprivileged, we can use the knet_handle internal change of this patch + transport change, and in the test suite, we can include internals.h and set knet_h->testing (or whatever flags) to 1 before configuring any transport sockets via link_set_config. No API change, users see no difference, me happy, everybody almost happy :-)

For the long term (aka 2.0), I am more happy to change the api to add bitmap flags at init time (that was added and removed a few times over time).

Otherwise another option that I am more open (to some degrees) to accept, instead of adding a new knet_handle_new, is to make it an API config call for example:

knet_handle_enable_unprivileged or knet_handle_enable_testing(void);

both, once enabled should be able to revert their effect and can/should only take effect if no links are configured. What makes me twitch is the idea of exposing those config options to the end user to be honest.

They feel like a good source of confusion and debugging nightmares if we endup with apps that are not running RT or with proper socket configs.

fabbione commented:
One more idea. What about using system capabilities? I am really not familiar with them but IIRC it is possible for example to delegate to a un-privileged user ability do to network configuration changes (at socket level).. was it SYSCAP_NET ? or something along those lines. I think we can allow knet to run unprivileged if the user has the right capabilities, tho we will need to figure out how the whole thing work.

wferi commented:
> Hey guys, I am not particularly excited about adding a whole new api call just for testing.

It isn't just for testing (although that's my prime concern now; but I've already implanted the first Debian package with an environmental hack to get some exposure), any non-performance-critical application could benefit, I think. Or ones needing other flags. Or maybe not. How was the value of `KNET_RING_RCVBUFF` determined?

> If we are a looking at a short term solution to allow test suite to run unprivileged, we can use the knet_handle internal change of this patch + transport change, and in the test suite, we can include internals.h and set knet_h->testing (or whatever flags) to 1 before configuring any transport sockets via link_set_config. No API change, users see no difference, me happy, everybody almost happy :-)

Unfortunately that won't work, because `knet_handle_new()` fails, thus we've got no `knet_h` to fiddle with. We'd have to delay socket setup to enable this route.

> Otherwise another option that I am more open (to some degrees) to accept, instead of adding a new knet_handle_new, is to make it an API config call for example:
> 
> knet_handle_enable_unprivileged or knet_handle_enable_testing(void);
> 
> both, once enabled should be able to revert their effect and can/should only take effect if no links are configured.

Do you mean setting a global variable, not a field in `knet_handle`? That would require a new mutex as well. I find that more complicated and less general than the flags suggestion.

> What makes me twitch is the idea of exposing those config options to the end user to be honest.
> They feel like a good source of confusion and debugging nightmares if we end up with apps that are not running RT or with proper socket configs.

I certainly don't understand the implications. Poor performance is one thing, misbehavior is a totally different one. The current proposal is clearly visible in the code **and** in the logs, beating my original solution using an environmental variable.

> One more idea. What about using system capabilities? I am really not familiar with them but IIRC it is possible for example to delegate to a un-privileged user ability do to network configuration changes (at socket level).. was it SYSCAP_NET ? or something along those lines. I think we can allow knet to run unprivileged if the user has the right capabilities, tho we will need to figure out how the whole thing work.

It's `CAP_NET_ADMIN`. Capabilities are slices of the full root privilege, thus a process with capabilities isn't really unprivileged. Setting realtime scheduling priorities requires `CAP_SYS_NICE` in addition. But if your launcher can hand out these capabilities to the knet application, it could also do the necessary operations itself before starting the knet application (if not that precisely targeted).

fabbione commented:
apparently my email didn´t make it to github (copy/paste and hopefully it won´t arrive late and duplicate my comments ;))

On 1/30/2018 11:40 PM, wferi wrote:
>
>     Hey guys, I am not particularly excited about adding a whole new api call just for testing.
>
> It isn't just for testing (although that's my prime concern now; but I've already implanted the first Debian package with an environmental hack to get some exposure), any non-performance-critical application could benefit, I think. Or ones needing other flags. Or maybe not. How was the value of KNET_RING_RCVBUFF determined?

KNET_RING_*BUFF were determined by heavy testing on both baremetal and virtual environments. Values lower or higher than that gives crappy performances.

>     If we are a looking at a short term solution to allow test suite to run unprivileged, we can use the knet_handle internal change of this patch + transport change, and in the test suite, we can include internals.h and set knet_h->testing (or whatever flags) to 1 before configuring any transport sockets via link_set_config. No API change, users see no difference, me happy, everybody almost happy :-)
>
> Unfortunately that won't work, because knet_handle_new() fails, thus we've got no knet_h to fiddle with. We'd have to delay socket setup to enable this route.

Oh right, we have the internal socketpair creation in knet_handle_new to account for. I forgot about those.

>     Otherwise another option that I am more open (to some degrees) to accept, instead of adding a new knet_handle_new, is to make it an API config call for example:
>
>     knet_handle_enable_unprivileged or knet_handle_enable_testing(void);
>
>     both, once enabled should be able to revert their effect and can/should only take effect if no links are configured.
>
> Do you mean setting a global variable, not a field in knet_handle? That would require a new mutex as well. I find that more complicated and less general than the flags suggestion.

Correct. We have the mutex in place already (see handle.c). The thing is really that I don´t really like the idea to change API already in 1.x series. The correct path is to enable flags as chrissie already suggested.

right now i am trying to compromise between 1.0 ability to run tests unprivileged without changing the API and 2.0 to make it clean with flags.

>     What makes me twitch is the idea of exposing those config options to the end user to be honest.
>     They feel like a good source of confusion and debugging nightmares if we end up with apps that are not running RT or with proper socket configs.
>
> I certainly don't understand the implications. Poor performance is one thing, misbehavior is a totally different one. The current proposal is clearly visible in the code and in the logs, beating my original solution using an environmental variable.

The implications are that we never tested knet as normal users or with buffers different than what´s set now. there might be scheduling issues that we are not aware of just yet. I think a change of this kind should first boil for a while in master, we will need to run CI as normal user for a long time before I would feel comfy to merge it down.

>     One more idea. What about using system capabilities? I am really not familiar with them but IIRC it is possible for example to delegate to a un-privileged user ability do to network configuration changes (at socket level).. was it SYSCAP_NET ? or something along those lines. I think we can allow knet to run unprivileged if the user has the right capabilities, tho we will need to figure out how the whole thing work.
>
> It's CAP_NET_ADMIN. Capabilities are slices of the full root privilege, thus a process with capabilities isn't really unprivileged. Setting realtime scheduling priorities requires CAP_SYS_NICE in addition. But if your launcher can hand out these capabilities to the knet application, it could also do the necessary operations itself before starting the knet application (if not that precisely targeted).

ok gotcha. doesn´t sound like something we want to use.

Fabio

fabbione commented:
took some time to think at it and you are both right, there is no clean way to do this without another API call.

I like better @chrissie-c idea to have a bitmap set of flags than just one boolean for priv/unpriv.

As for the API call name, knet_handle_new_extended or knet_handle_new_opts perhaps? we need something more generic than just _privileged.

Please remember to bump the library minor version to reflect adding an API call.

wferi committed:
New API function: knet_handle_new_ex() taking handle flags

For now, the only supported flag is KNET_HANDLE_FLAG_PRIVILEGED,
which is also the default in knet_handle_new().

The tests are adapted to switch this flag off, which makes them pass
in default environments without privileges.

Signed-off-by: Ferenc Wágner <wferi@debian.org>

wferi commented:
So, here it is again with new names and bitmap flags.
I think bumping the library version belongs to a different commit right before release, when ABI compatibility status is evaluated. But this is certainly question of policy.

fabbione commented:
On 2/18/2018 7:07 PM, wferi wrote:
>
> So, here it is again with new names and bitmap flags.
> I think bumping the library version belongs to a different commit
> right before release, when ABI compatibility status is evaluated. But
> this is certainly question of policy.
>

We didn´t establish any specific policy on this subject so we can just
agree what is best for us.

> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kronosnet/kronosnet/pull/124#issuecomment-366534778>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAW6rQ4O88hdGG1Z7HIpFjjtGUmhbAnfks5tWGbQgaJpZM4RyQml>.
>

fabbione commented:
I am looking at the TODO list for 1.1, we only have this change and moving man pages generation to top level dir, then we can release.You can just push a commit to change library soname within this PR.

fabbione commented:
oh and btw, the patch looks good

wferi committed:
Indicate new API call in libtool version info

Signed-off-by: Ferenc Wágner <wferi@debian.org>

wferi committed:
build: provide -version-info directly instead of -version-number

The info components carry the real OS-independent library version
information, from which the version number is computed differently
on different systems. According to the libtool manual "new projects
should use the -version-info flag instead".

Signed-off-by: Ferenc Wágner <wferi@debian.org>

wferi commented:
Here you go. I don't really understand the master/stable1/stable1-proposed distinction at the moment, because I can't see anything on master which isn't 1.1 material, so we could pretty much release 1.1 straight from master. If we do so, this PR can go into master in full. If not, then the version info bump should probably not go there (as recommended by the second point of [Libtool manual 7.3](https://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html)).

wferi commented:
I forgot installing the new man page, thus the fixup. It should be folded before merge, but I didn't want to rebase this PR during review.

fabbione commented on libknet/Makefile.am:
```diff
@@ -86,7 +86,7 @@ EXTRA_libknet_la_DEPENDENCIES	= $(SYMFILE)
 			  -Wl,--version-script=$(srcdir)/$(SYMFILE) \
 			  -Wl,-rpath=$(pkglibdir) \
-			  -version-number $(libversion)
+			  -version-info $(libversion)
```
why using info vs number? what´s the benefit? I find it a lot harder to read vs number that matches soname installed by the file.

wferi replied:
The info components carry the real OS-independent library version information. The version number is computed differently on different systems. Look for "vinfo_number" and "Calculate the version variables" in the generated `libtool` script. The [libtool manual](https://www.gnu.org/software/libtool/manual/html_node/Link-mode.html) also says that "new projects should use the -version-info flag instead".

fabbione commented on libknet/Makefile.am:
```diff
@@ -17,7 +17,7 @@ EXTRA_DIST		= $(SYMFILE)
 SUBDIRS			= . tests man
 
-libversion		= 1:0:0
+libversion		= 2:0:1
```
what´s the equivalent soname for this one? in theory adding an API should bump to so.1.1.0 IIRC. we are not break backward compatibility.

wferi replied:
Yes, and that's exactly what happens in this case.

fabbione replied:
ok, i guess i´ll need to get used to "weird" numbers ;)

I am going to merge https://github.com/kronosnet/kronosnet/pull/136 and then we can rebase this branch on top of master and ack/merge.

wferi replied:
They are weird indeed. I've seen several projects quote (or at least link to) the libtool manual in a comment right above their version info... and to be honest I'm inclined to do the same.

fabbione commented:
On 2/19/2018 10:43 AM, wferi wrote:
>
> Here you go. I don't really understand the
> master/stable1/stable1-proposed distinction at the moment, because I
> can't see anything on master which isn't 1.1 material, so we could
> pretty much release 1.1 straight from master.
>

https://lists.kronosnet.org/pipermail/devel/2018-January/000049.html

you are right, at the moment master and stable1-* haven´t diverged yet
but they will in future .

Fabio

> If we do so, this PR can go into master in full. If not, then the
> version info bump should probably not go there (as recommended by the
> second point of Libtool manual 7.3
> <https://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html>).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kronosnet/kronosnet/pull/124#issuecomment-366636378>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAW6rfg1fp-tctvvcpOtcafc6PJxuRvqks5tWUJRgaJpZM4RyQml>.
>

fabbione commented:
On 2/19/2018 4:59 PM, wferi wrote:
>
> *@wferi* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In libknet/Makefile.am
> <https://github.com/kronosnet/kronosnet/pull/124#discussion_r169115508>:
>
> > @@ -17,7 +17,7 @@ EXTRA_DIST		= $(SYMFILE)
>  
>  SUBDIRS			= . tests man
>  
> -libversion		= 1:0:0
> +libversion		= 2:0:1
>
> They are weird indeed. I've seen several projects quote (or at least
> link to) the libtool manual in a comment right above their version
> info... and to be honest I'm inclined to do the same.
>

I´ll never object to links to docs :-)

> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kronosnet/kronosnet/pull/124#discussion_r169115508>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAW6rRGjfAjSCy8V7uxwvFugkr5EMiK9ks5tWZpbgaJpZM4RyQml>.
>

wferi commented:
I'm getting uncertain whether 64 flag bits are really justified... Shall I go back to a lesser type to reduce waste?

fabbione commented:
On 2/20/2018 8:19 AM, wferi wrote:
>
> I'm getting uncertain whether 64 flag bits are really justified...
> Shall I go back to a lesser type to reduce waste?
>

Nah, let´s keep the flexibility there. 32bit are not expensive :-)

Fabio

> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kronosnet/kronosnet/pull/124#issuecomment-366887419>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAW6rSPz4_bU5Lj8y4bbQLP3ffWG08lyks5tWnIIgaJpZM4RyQml>.
>
